File: app/src/main/java/com/android/example/popularmovies/ui/detail/DetailViewModel.java
package com.android.example.popularmovies.ui.detail;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.android.example.popularmovies.AppExecutors;
import com.android.example.popularmovies.database.FavMovieDatabase;
import com.android.example.popularmovies.database.MovieEntry;

import java.util.Objects;

public class DetailViewModel extends AndroidViewModel {

    private MutableLiveData<MovieEntry> movie;
    private MutableLiveData<Boolean> isFav;

    public LiveData<MovieEntry> getMovie() {
        return movie;
    }

    public LiveData<Boolean> getIsFav() {
        return isFav;
    }

    public DetailViewModel(Application app, MovieEntry movie){
        super(app);
        this.movie = new MutableLiveData<>();
        this.movie.setValue(movie);
        isFav = new MutableLiveData<>();
        initializeIsFav();
    }

    private boolean isMovieFavorite(){
        int id = Objects.requireNonNull(movie.getValue()).getId();
        return FavMovieDatabase.getInstance(getApplication()).favMovieDao()
                .countMovieById(id) > 0;
    }

    private void initializeIsFav(){
        AppExecutors.getInstance().diskIO().execute(() -> {
            int id = Objects.requireNonNull(movie.getValue()).getId();
            isFav.postValue(FavMovieDatabase.getInstance(getApplication()).favMovieDao()
                    .countMovieById(id) > 0);
        });
    }

    public synchronized void addToFavs(){
        if (isFav.getValue() != null && !isFav.getValue()){
            AppExecutors.getInstance().diskIO().execute(() -> {
                FavMovieDatabase.getInstance(getApplication()).favMovieDao()
                        .addMovieToFavorites(movie.getValue());
                isFav.postValue(true);
            });
        }
    }

    public synchronized void removeFromFavs(){
        if (isFav.getValue() != null && isFav.getValue()){
            AppExecutors.getInstance().diskIO().execute(() -> {
                FavMovieDatabase.getInstance(getApplication()).favMovieDao()
                        .removeMovieFromFavorites(movie.getValue());
                isFav.postValue(false);
            });
        }
    }
}

File: app/src/main/java/com/android/example/popularmovies/ui/detail/DetailViewModel.kt
package com.android.example.popularmovies.ui.detail

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.android.example.popularmovies.database.FavMovieDatabase
import com.android.example.popularmovies.database.MovieEntry
import kotlinx.coroutines.launch

class DetailViewModel(private val app: Application, movie: MovieEntry) : AndroidViewModel(app) {
    private val _movie: MutableLiveData<MovieEntry> = MutableLiveData(movie)
    val movie: LiveData<MovieEntry>
        get() = _movie
    private val _isFav: MutableLiveData<Boolean> = MutableLiveData()
    val isFav: LiveData<Boolean>
        get() = _isFav

    private val isMovieFavorite: Boolean
        get() {
            val id = requireNotNull(movie.value).id
            return FavMovieDatabase.getInstance(app)!!.favMovieDao()
                    .countMovieById(id) > 0
        }

    private fun initializeIsFav() {
        viewModelScope.launch {
            val id = requireNotNull(movie.value).id
            _isFav.postValue(requireNotNull(FavMovieDatabase.getInstance(requireNotNull(app.applicationContext))
                    ?.favMovieDao()?.countMovieById(id)) > 0)
        }
        /*AppExecutors.Companion.getInstance().diskIO().execute(Runnable {
            val id = Objects.requireNonNull(_movie.value).getId()
            isFav.postValue(FavMovieDatabase.Companion.getInstance(getApplication<Application>())!!.favMovieDao()
                    .countMovieById(id) > 0)
        })*/
    }

    @Synchronized
    fun addToFavs() {
        if (_isFav.value != null && !_isFav.value!!) {
            viewModelScope.launch {
                FavMovieDatabase.getInstance(app.applicationContext)!!.favMovieDao()
                        .addMovieToFavorites(movie.value!!)
                _isFav.postValue(true)
            }
        }
    }

    @Synchronized
    fun removeFromFavs() {
        if (_isFav.value != null && _isFav.value!!) {
            viewModelScope.launch {
                FavMovieDatabase.getInstance(app.applicationContext)!!.favMovieDao()
                        .removeMovieFromFavorites(movie.value!!)
                _isFav.postValue(false)
            }
        }
    }

    init {
        initializeIsFav()
    }
}